control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -2986,74 +2987,171 @@ func (r *HostedControlPlaneReconciler) reconcileIngressOperator(ctx context.Cont
+		},
+		{
+			name:       "redHatOperatorsService",
+			manifest:   manifests.RedHatOperatorsService(hcp.Namespace),
+			reconciler: olm.ReconcileRedHatOperatorsService,
+		},
+	}
+
+	// Check if the defaultSources are disabled in the HostedCluster's OperatorHub object
+	operatorHub := &configv1.OperatorHub{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+	}
+
+	// Creating a new client for the HostedCluster to check the OperatorHub object
+	hcClient, err := generateHostedClusterClient(ctx, r.Client, hcp.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed to get Hosted Cluster client: %w", err)
+	}
+
+	if err := hcClient.Get(ctx, client.ObjectKeyFromObject(operatorHub), operatorHub); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get OperatorHub %s: %w", client.ObjectKeyFromObject(operatorHub).String(), err)
+		}
+	}
+
 	if hcp.Spec.OLMCatalogPlacement == hyperv1.ManagementOLMCatalogPlacement {
-		overrideImages, err := checkCatalogImageOverides(p.CertifiedOperatorsCatalogImageOverride, p.CommunityOperatorsCatalogImageOverride, p.RedHatMarketplaceCatalogImageOverride, p.RedHatOperatorsCatalogImageOverride)
-		if err != nil {
-			return fmt.Errorf("failed to reconcile catalogs: %w", err)
+
+		// Clean HostedCluster OLM pods
+		olmLabelKey := "olm.catalogSource"
+
+		pods := &corev1.PodList{}
+		if err := hcClient.List(ctx, pods, &client.ListOptions{
+			Namespace: "openshift-marketplace",
+		}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to get pods from openshift-marketplace namespace: %w", err)
+			}
 		}
 
-		catalogsImageStream := manifests.CatalogsImageStream(hcp.Namespace)
-		if !overrideImages {
-			isImageRegistryOverrides := util.ConvertImageRegistryOverrideStringToMap(p.OLMCatalogsISRegistryOverridesAnnotation)
-			if _, err := createOrUpdate(ctx, r, catalogsImageStream, func() error {
-				return olm.ReconcileCatalogsImageStream(catalogsImageStream, p.OwnerRef, isImageRegistryOverrides)
-			}); err != nil {
-				return fmt.Errorf("failed to reconcile catalogs image stream: %w", err)
+		for _, pod := range pods.Items {
+			labelValue, labelExists := pod.Labels[olmLabelKey]
+
+			if labelExists && (labelValue == "community-operators" ||
+				labelValue == "redhat-operators" ||
+				labelValue == "certified-operators" ||
+				labelValue == "redhat-marketplace") {
+				if err := hcClient.Delete(ctx, &pod); err != nil {
+					if !apierrors.IsNotFound(err) {
+						return fmt.Errorf("failed to delete pod %s from openshift-marketplace namespace: %w", pod.Name, err)
+					}
+				}
+			}
+		}
+
+		if operatorHub.Spec.DisableAllDefaultSources {
+			for _, svc := range olmSvcReconciler {
+				if err := r.Client.Delete(ctx, svc.manifest); err != nil {
+					if !apierrors.IsNotFound(err) {
+						return fmt.Errorf("failed to delete %s service on OLM reconcile: %w", svc.name, err)
+					}
+				}
+			}
+
+			for _, dep := range olmDepReconciler {
```

Comment:
Why do we want to do this? what happens if we don't? We already delete the catalogs in the hcco. We wouldn't want to expose anything in the guest cluster that directly impacts the footprint of what we run in the management side

Reply:
But in HCCO we are deleting the deployments and services in the data plane. When placement is set to `management` it does not delete the same objects from the HCP side.

Reply:
> Why do we want to do this?

Basically when you change the operatorHub in the HC, disabling the defaultSources, when we are working in guest mode, it behaves properly and deletes the deployment and services. When we work in `management` mode, the deployment and services stays there.

> We already delete the catalogs in the hcco

Yes but using the hcClient, not using the cpClient. I was thinking that the HCCO cannot perform that kind of destructive actions in the HCP side.

If we can, the implementation will be very different, doing all the things in the HCCO and only changing the client we use in the services and deployment deletion. Makes sense?

Reply:
>Basically when you change the operatorHub in the HC, disabling the defaultSources, when we are working in guest mode, it behaves properly and deletes the deployment and services. When we work in management mode, the deployment and services stays there.

Thanks! I see that's what the PR is doing but I'm wondering why we need it. What happens if we don't do it (as now)?
This PR is exposing and API in the guest cluster that would enable consumers to delete things in the management cluster which we don't want to.

>We already delete the catalogs in the hcco

What I mean with this is that we already provide consistent UX to the guest cluster user by deleting those things.

Reply:
> What happens if we don't do it (as now)?

(as it's now) If we disable de operatorHub's default sources and the placement is set in `management` mode, the  OLM related pods will keep there. That impacts mostly for self-manage disconnected because the pods will remain there as "ImagePullBackOff".